File: ClassAnalyzer.py
# This Python file uses the following encoding: utf-8

import os
import re
import concurrent.futures

from specialvals import DEBUGMODE


class ClassAnalyzer:
    def __init__(self):
        pass


    def _findClassAttributesExternal(self, cls, classes):
        class_attributes = dict()

        myFuncs = [_ for _ in classes[cls]]

        for clsx in classes:
            if clsx != cls:
                for func in classes[clsx]:
                    active_attribs = []
                    for line in classes[clsx][func]:
                        tmo = line.rstrip("\n")
                        if tmo.find("//") >= 0:
                            tmo = tmo[0:tmo.index("//")]

                        if cls in line:
                            regex = '( '+cls+'[\*]* [a-z0-9_]+ = )'
                            mo = re.search(regex, tmo)
                            mo2 = re.search('([a-z0-9_]+\-\>[a-zA-Z_<>0-9\*]+\()', tmo)

                            nomo = ""
                            if mo:
                                tx = mo.group().strip().rstrip('=').rstrip()
                                tx = tx.split(' ')[1]
                                active_attribs.append(tx)
                                nomo = '"' + tx + '";"' + tmo.strip().rstrip(';')
                            elif mo2:
                                tx = mo2.group().strip()
                                txa = tx[tx.index('>') + 1:]
                                txa = txa[0:tx.index('(')-1]
                                if txa in myFuncs:
                                    nomo = '"' + tx.split('(')[0] + '";"' + tmo.strip().rstrip(';')

                            if len(nomo) > 0:
                                nomo += '";"' + clsx + '";"' + func + '"\n'
                                key = nomo.split(';')[0]
                                if not key in class_attributes:
                                    class_attributes[key] = nomo
                            else:
                                for frag in active_attribs:
                                    self._benulf(line, classes, class_attributes, clsx, cls, func, fragger=frag)
                        else:
                            for frag in active_attribs:
                                self._benulf(line, classes, class_attributes, clsx, cls, func, fragger=frag)

        # TODO: remove these later when attribs found!
        with open(cls + ".endl", "w") as f:
            for attr in class_attributes:
                f.write(class_attributes[attr])

        return [class_attributes[_] for _ in class_attributes]


    def _benulf(self, line, classes, class_attributes, cls, orgCls, func, fragger="rdi"):
        tmo = line.rstrip("\n")
        if tmo.find("//") >= 0:
            tmo = tmo[0:tmo.index("//")]

        mo = re.search("([\*]*[\(]*\("+fragger+"[ +]*[^a-zA-Z0-9]+[ \+\*\[\]\(\)x0-9a-zA-Z]+\)[\)]*)", tmo)
        mo2 = re.search("([ ]"+fragger+"[ +]*[ \;][ \+\*\[\]\(\)x0-9a-zA-Z=]*)", tmo)
        mo3 = re.search("(STRUCT_[0-9\*]+ [a-z0-9_]+ = "+fragger+"[;])", tmo)

        nomo = ""
        if mo:
            tx = self._clamsFix(tmo, mo.group())
            nomo = '"' + tx + '";"' + tmo.strip().rstrip(';')
        elif mo2:
            tx = self._clamsFix(tmo, mo2.group())
            nomo = '"' + tx + '";"' + tmo.strip().rstrip(';')
        elif fragger + "*" in tmo or fragger + " " in tmo or (fragger == "rdi" and "this->" in tmo and not tmo[tmo.index("this->") + 6:].split("(")[0] in classes[cls]):
            nomo = '"' + fragger + '";"' + tmo.strip().rstrip(';')

        if len(nomo) > 0:
            nomo += '";"' + cls + '";"' + func + '"\n'
            key = nomo.split(';')[0]
            if not key in class_attributes:
                class_attributes[key] = nomo

        if mo3:
            xa = tmo[tmo.index("STRUCT_"):]
            xa = xa[0:xa.index(" ")]
            nomo = '"' + xa + '";"' + tmo.strip().rstrip(';') + '";"' + cls + '";"' + func + '"\n'
            key = nomo.split(';')[0]
            if not key in class_attributes:
                class_attributes[key] = nomo


    def _findClassAttributesS(self, cls, classes):
        class_attributes = dict()
        for func in classes[cls]:
            for line in classes[cls][func]:
                self._benulf(line, classes, class_attributes, cls, cls, func)

        with open(cls + ".endl", "w") as f:
            for attr in class_attributes:
                f.write(class_attributes[attr])

        return [class_attributes[_] for _ in class_attributes]


    def _clamsFix(self, allText, text):
        clam = 0
        lok = False
        for c in text:
            if c == "(":
                clam += 1
                lok = True
            elif c == ")":
                clam -= 1
        tx = text.rstrip("\n").strip()
        xon = allText[allText.index(tx) + len(tx):]
        for c in xon:
            if c == "(":
                clam += 1
            elif c == ")":
                clam -= 1
            tx += c
            if clam <= 0:
                break
        while clam < 0:
            rev_tx = tx[::-1]
            lastIndex = len(tx) - rev_tx.index(")") - 1
            tx = tx[0:lastIndex]
            clam += 1
        tx = tx.rstrip(';')
        lok = lok and tx[len(tx) - 1] != ")"
        if lok:
            rev_tx = tx[::-1]
            x = 0
            for c in rev_tx:
                if c == ")":
                    break
                x += 1
            tx = tx[0:len(tx) - x]
        return tx


    def _help_findClassAttributes(self, classes, partNum, maxPartCount):
        packSize = int(len(classes) / maxPartCount)
        startIndex = int(packSize * partNum)
        endIndex = startIndex + packSize
        if partNum + 1 == maxPartCount:
            endIndex = len(classes.keys())
        keyList = list(classes.keys())[startIndex:endIndex]
        class_attributes = dict()
        for cls in keyList:
            class_attributes[cls] = self._findClassAttributesS(cls, classes)
            class_attributes[cls].extend(self._findClassAttributesExternal(cls, classes))
        #print("Process", partNum, "finished!")
        return [len(keyList), class_attributes, os.getpid()]


    def findClassAttributes(self, classes):
        print("Analyzing - adding class attributes...")

        maxProcs = int(os.cpu_count() * 0.75) # 16 -> 12, 8 -> 6, 4 -> 3
        with concurrent.futures.ProcessPoolExecutor() as executor:
            results = [executor.submit(self._help_findClassAttributes, classes, i, maxProcs) for i in range(maxProcs)]

        values_list = []
        for f in concurrent.futures.as_completed(results):
            values = f.result()
            values_list.append(values)

        class_attributes = dict()
        for vx in values_list:
            for fc in vx[1]:
                class_attributes[fc] = vx[1][fc]
            #print("Processed", vx[0], "classes with process", vx[2])

        print("Analyzing - adding class attributes...DONE")

        return class_attributes

File: cpp_proj/classanalyzer.cpp
#include "classanalyzer.h"

ClassAnalyzer::ClassAnalyzer()
{
  QString exDir(QString("./endl/class_info").replace("/", QDir::separator()));
  if (!QDir().mkpath(exDir)) {
      cout << "Could not create " << exDir.toStdString().c_str() << "!" << endl;
  }
}

QString ClassAnalyzer::clamsFix(QString allText, QString text)
{
  int clam = 0;
  bool lok = false;

  for (int i = 0; i < text.length(); i++) {
      QChar c = text.at(i);
      if (c == '('){
          clam++;
          lok = true;
      } else if (c == ')') {
          clam--;
      }
  }

  QString tx = text.remove(QRegExp("([\n]+$)")).trimmed();
  QString xon = allText.mid(allText.indexOf(tx) + tx.length());

  for (int i = 0; i < xon.length(); i++) {
      QChar c = xon.at(i);
      if (c == '('){
          clam++;
      } else if (c == ')') {
          clam--;
      }

      tx += c;
      if (clam <= 0) break;
  }

  QString revTx;
  while (clam < 0) {
      tx.reserve(tx.length());
      revTx = QString(tx);
      tx.reserve(tx.length());
      int lastIndex = tx.length() - revTx.indexOf(")") - 1;
      tx = tx.left(lastIndex);
      clam++;
  }

  tx = tx.remove(QRegExp("([;]$)"));
  lok = (lok && tx.at(tx.length() - 1) != ')');
  if (lok) {
      tx.reserve(tx.length());
      revTx = QString(tx);
      tx.reserve(tx.length());
      int x = 0;

      for (int i = 0; i < revTx.length(); i++) {
          QChar c = revTx.at(i);
          if (c == ')'){
              break;
          }
          x++;
      }

      tx = tx.left(tx.length() - x);
  }

  return tx;
}

void ClassAnalyzer::benulf(QString line, map<QString, FixedClass>* classes, map<QString, QString> *classAttributes, QString cls, QString orgCls, QString func, QString fragger)
{
  QString tmo = line.remove(QRegExp("([\n]+$)"));
  if (tmo.contains("//")) {
      tmo = tmo.left(tmo.indexOf("//"));
  }

  QRegExp mo("([*]*[\(]*\("+fragger+"[ +]*[^a-zA-Z0-9]+[ [+][*]\\[\\]\\(\\)x0-9a-zA-Z]+\\)[\\)]*)");
  mo.indexIn(tmo);

  QRegExp mo2("([ ]"+fragger+"[ +]*[ \\;][ [+][*]\\[\\]\\(\\)x0-9a-zA-Z=]*)");
  mo2.indexIn(tmo);

  QRegExp mo3("(STRUCT_[0-9*]+ [a-z0-9_]+ [=] "+fragger+"[;])");
  mo3.indexIn(tmo);

  QString tx;
  QString nomo("");
  if (mo.captureCount() > 0) {
      tx = this->clamsFix(tmo, mo.cap());
      nomo = "\"" + tx + "\";\"" + tmo.trimmed().remove(QRegExp("([;]+$)"));
  } else if (mo2.captureCount() > 0) {
      tx = this->clamsFix(tmo, mo2.cap());
      nomo = "\"" + tx + "\";\"" + tmo.trimmed().remove(QRegExp("([;]+$)"));
  } else if (tmo.contains(fragger + "*") ||
             tmo.contains(fragger + " ") ||
             (fragger == QString("rdi") && tmo.contains("this->"))) {
      bool ok = true;
      if (fragger == QString("rdi") && tmo.contains("this->")) {
        QString searchX = tmo.mid(tmo.indexOf("this->") + 6).split('(').first();
        foreach (FixedFunction f, classes->at(cls).getFunctions()) {
            if (f.getName() == searchX) {
                ok = false;
            }
        }
      }
      if (ok) {
          nomo = "\"" + fragger + "\";\"" + tmo.trimmed().remove(QRegExp("([;]$)"));
      }
  }

  if (nomo.length() > 0) {
      nomo += "\";\"" + cls + "\";\"" + func + "\"\n";
      QString key = nomo.split(';').first();
      if (classAttributes->count(key) == 0) {
          classAttributes->insert_or_assign(key, nomo);
      }
  }

  if (mo3.captureCount() > 0) {
      QString xa = tmo.mid(tmo.indexOf("STRUCT_"));
      xa = xa.left(xa.indexOf(" "));
      nomo = "\"" + xa + "\";\"" + tmo.trimmed().remove(QRegExp("([;]+$)")) + "\";\"" + cls + "\";\"" + func + "\"\n";
      QString key = nomo.split(';').first();
      if (classAttributes->count(key) == 0) {
          classAttributes->insert_or_assign(key, nomo);
      }
  }
}

QStringList ClassAnalyzer::findClassAttributesExternal(QString cls, map<QString, FixedClass> *classes)
{
  map<QString, QString> classAttributes;
  vector<FixedFunction> myFuncs = classes->at(cls).getFunctions();

  foreach (auto clsx, *classes) {
      if (clsx.first == cls) continue; // skip same class

      foreach (FixedFunction func, clsx.second.getFunctions()) {
          QStringList activeAttribs;
          foreach (QString line, func.getCodeLines()) {
              QString tmo = line.remove(QRegExp("([\n]+$)"));
              if (tmo.contains("//")) {
                  tmo = tmo.left(tmo.indexOf("//"));
              }

              if (tmo.contains(cls)) {
                  QRegExp moa("( "+cls+"[*]* [a-z0-9_]+ = )");
                  int moaPos = moa.indexIn(tmo);

                  QRegExp mob("( "+cls+"[*]* [a-z0-9_]+[;])");
                  int mobPos = mob.indexIn(tmo);

                  QRegExp mo2("([a-z0-9_]+[-][>][a-zA-Z_<>0-9*]+\()");
                  int mo2Pos = mo2.indexIn(tmo);

                  QString tx;
                  QString nomo("");
                  if (moaPos >= 0 || mobPos >= 0) {
                      if (moaPos >= 0) {
                          QStringList mx = moa.capturedTexts();
                          tx = mx.first().trimmed().remove(QRegExp("([=]+$)")).remove(QRegExp("([ ]+$)"));
                      } else {
                          QStringList mx = mob.capturedTexts();
                          tx = mx.first().trimmed().remove(QRegExp("([;]+$)"));
                      }
                      tx = tx.split(' ')[1];
                      activeAttribs.append(tx);
                      nomo = "\"" + tx + "\";\"" + tmo.trimmed().remove(QRegExp("([;]+$)"));
                  } else if (mo2Pos >= 0) {
                      tx = mo2.cap().trimmed();
                      QString txa = tx.mid(tx.indexOf('>') + 1);
                      txa = txa.left(txa.indexOf('(')- 1); // FIXME: possible inde bug because txa was tx before in python
                      foreach (auto f, myFuncs) {
                          if (f.getName() == txa) {
                              nomo = "\"" + tx.split('(').first() + "\";\"" + tmo.trimmed().remove(QRegExp("([;]+$)"));
                              break;
                          }
                      }
                  }

                  if (nomo.length() > 0) {
                      nomo += "\";\"" + clsx.first + "\";\"" + func.getName() + "\"\n";
                      QString key = nomo.split(';').first();
                      if (classAttributes.count(key) == 0) {
                          classAttributes.insert_or_assign(key, nomo);
                      }
                  } else {
                      foreach (QString frag, activeAttribs) {
                          this->benulf(line, classes, &classAttributes, clsx.first, cls, func.getName(), frag);
                      }
                  }
              } else {
                  foreach (QString frag, activeAttribs) {
                      this->benulf(line, classes, &classAttributes, clsx.first, cls, func.getName(), frag);
                  }
              }
          }
      }
  }

  // TODO: remove these later when attribs found!
  QFile f(QString("./endl/" + cls + ".endl").replace("/", QDir::separator()));
  if (f.open(QFile::WriteOnly | QFile::Append | QFile::Text)) {
      QTextStream out(&f);
      foreach (auto attr, classAttributes) {
          out << attr.second;
      }
      f.close();
  }

  QStringList ret;
  foreach (auto attr, classAttributes) {
      ret.append(attr.second);
  }

  return ret;
}

QStringList ClassAnalyzer::findClassAttributes(QString cls, map<QString, FixedClass> *classes)
{
  map<QString, QString> classAttributes;
  foreach (FixedFunction func, classes->at(cls).getFunctions()) {
      foreach (QString line, func.getCodeLines()) {
          this->benulf(line, classes, &classAttributes, cls, cls, func.getName());
      }
  }

  QFile f(QString("./endl/" + cls + ".endl").replace("/", QDir::separator()));
  if (f.open(QFile::WriteOnly | QFile::Append | QFile::Text)) {
      QTextStream out(&f);
      foreach (auto attr, classAttributes) {
          out << attr.second << Qt::endl;
      }
      f.close();
  }

  QStringList ret;
  foreach (auto attr, classAttributes) {
      ret.append(attr.second);
  }
  return ret;
}

map<QString, QStringList> ClassAnalyzer::findClassAttributes(map<QString, FixedClass> *classes)
{
  map<QString, QStringList> classAttributes;

  cout << "Add used class imports..." << endl;

  uint processor_count = processorCount();
  //processor_count = 1; // for testing bugs
  int length = classes->size() / processor_count;
  for (uint i = 0; i < processor_count; i++) {
      int start = i * length;
      int end = start + length;
      if (i == processor_count - 1)
          end = classes->size();

      ClassAnalyzerTask *hello = new ClassAnalyzerTask(this, &classAttributes, classes, start, end);
      // QThreadPool takes ownership and deletes 'hello' automatically
      QThreadPool::globalInstance()->start(hello);
  }
  QThreadPool::globalInstance()->waitForDone(); // waits for all to be done(?)

  cout << "DONE" << endl;

  return classAttributes;
}
